resource/resource.go

```diff
@@ -342,6 +342,25 @@ func (r *Resource) find(ctx context.Context, q *query.Query, forceTotal bool) (l
 	return
 }
 
+// Count implements Storer interface.
+func (r *Resource) Count(ctx context.Context, q *query.Query) (n int, err error) {
+	if LoggerLevel <= LogLevelDebug && Logger != nil {
+		defer func(t time.Time) {
+			found := n
+			Logger(ctx, LogLevelDebug, fmt.Sprintf("%s.Count(...)", r.path), map[string]interface{}{
+				"duration": time.Since(t),
+				"found":    found,
+				"error":    err,
+			})
+		}(time.Now())
+	}
+	if err = r.hooks.onFind(ctx, q); err == nil {
+		n, err = r.storage.Count(ctx, &query.Query{Predicate: q.Predicate})
```

Comment:
How can this compile? I don’t see where you are adding `Count` to the storage type.

Reply:
`r.storage` is defined here:
```
type Resource struct {
...
	storage     storageHandler
...
}

type storageHandler interface {
	Storer
	MultiGetter
	Counter
	Get(ctx context.Context, id interface{}) (item *Item, err error)
}
```
But
```
type storageWrapper struct {
	Storer
}
```
Satisfies `storageHandler` and has `Count` function:
```
func (s storageWrapper) Count(ctx context.Context, q *query.Query) (total int, err error) {
	if s.Storer == nil {
		return -1, ErrNoStorage
	}
	if ctx.Err() != nil {
		return -1, ctx.Err()
	}
	if c, ok := s.Storer.(Counter); ok {
		return c.Count(ctx, q)
	}
	return -1, ErrNotImplemented
}
```

Reply:
With the implementations you list though, should we not fallback to use Find with `Window{Limit=0}` when the Count interface is _not_ implemented, instead of returning an error?

Reply:
@smyrman This could work, lets see what @rs will say.

Reply:
@rs  any comment on this?

Reply:
Sounds good.